### PR TITLE
predict() support for pose estimation models

### PR DESCRIPTION
**Pull request by BloodAxe**
```
from super_gradients.common.object_names import Models
from super_gradients.training import models
from super_gradients.training.models import DEKRPoseEstimationModel

dekr: DEKRPoseEstimationModel = models.get(Models.DEKR_W32_NO_DC, pretrained_weights="coco_pose")
result = dekr.predict("Football-2.jpeg", conf=0.2)
result.show(keypoint_radius=8, joint_thickness=4)
```
pred_0

### CODE REVIEW

File: src/super_gradients/training/models/pose_estimation_models/dekr_hrnet.py [hunk 1]
```diff
@@ -520,8 +530,102 @@ def init_weights(self):
...
+        :param iou:             (Optional) IoU threshold for the nms algorithm
+        :param conf:            (Optional) Below the confidence threshold, prediction are discarded
+        """
+        self._joint_links = joint_links or self._joint_links
+        self._joint_colors = joint_colors or self._joint_colors
+        self._keypoint_colors = keypoint_colors or self._keypoint_colors
+        self._image_processor = image_processor or self._image_processor
+        self._default_nms_conf = conf or self._default_nms_conf
+
+    @lru_cache(maxsize=1)
```
Comment (ofrimasad) [inline src/super_gradients/training/models/pose_estimation_models/dekr_hrnet.py:567]: what is this for?

File: src/super_gradients/training/models/pose_estimation_models/dekr_hrnet.py [hunk 2]
```diff
@@ -520,8 +530,102 @@ def init_weights(self):
                 if hasattr(m, "bias"):
                     nn.init.constant_(m.translation_conv.bias, 0)
 
+    @staticmethod
+    def get_post_prediction_callback(conf: float = 0.05):
```
Comment (ofrimasad) [inline src/super_gradients/training/models/pose_estimation_models/dekr_hrnet.py:534]: question: don't you think we should maybe inherit `DEKRPoseEstimationModel` from `PoseEstimationModel` and move this logic to `PoseEstimationModel`?

File: src/super_gradients/training/models/prediction_pose_estimation_results.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+
+from super_gradients.training.models.prediction_results import ImagePrediction, ImagesPredictions, VideoPredictions
+from super_gradients.training.models.predictions import PoseEstimationPrediction
+from super_gradients.training.utils.media.image import show_image, save_image
+from super_gradients.training.utils.media.video import show_video_from_frames, save_video
+from super_gradients.training.utils.visualization.poses import draw_skeleton
+
+
+@dataclass
+class ImagePoseEstimationPrediction(ImagePrediction):
```
Comment (ofrimasad) [inline src/super_gradients/training/models/prediction_pose_estimation_results.py:15]: why is this file under the `models` directory? 
It looks like it belongs in some utils dir

File: src/super_gradients/training/models/prediction_pose_estimation_results.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+        :param joint_colors:    Optional list of tuples representing the colors for each joint.
+                                If None, default colors are used.
+                                If not None the length must be equal to the number of joint links in the skeleton.
+        :param joint_thickness: Thickness of the joint links  (in pixels).
+        :param keypoint_colors: Optional list of tuples representing the colors for each keypoint.
+                                If None, default colors are used.
+                                If not None the length must be equal to the number of joints in the skeleton.
+        :param keypoint_radius: Radius of the keypoints (in pixels).
+        :param show_confidence: Whether to show confidence scores on the image.
+        :param box_thickness:   Thickness of bounding boxes.
```
Comment (ofrimasad) [inline src/super_gradients/training/models/prediction_pose_estimation_results.py:46]: can we also drow without boxes?

File: src/super_gradients/training/models/predictions.py [hunk 5]
```diff
@@ -53,3 +53,56 @@ def _validate_input(self, bboxes: np.ndarray, confidence: np.ndarray, labels: np
 
     def __len__(self):
         return len(self.bboxes_xyxy)
+
+
+@dataclass
+class PoseEstimationPrediction(Prediction):
```
Comment (ofrimasad) [inline src/super_gradients/training/models/predictions.py:59]: this file as well

File: src/super_gradients/training/utils/visualization/poses.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
+from typing import Union, List, Tuple
+
+import cv2
+import numpy as np
+
+from super_gradients.training.utils.visualization.detection import draw_bbox
+
+
+def draw_skeleton(
```
Comment (ofrimasad) [inline src/super_gradients/training/utils/visualization/poses.py:9]: is the file name a mistake? (poses)

**Review by ofrimasad — COMMENTED** [review]
some minor comments and questions.
Are the colors defined here related to something? can't we just use some kind of colo-map?

File: src/super_gradients/recipes/dataset_params/coco_pose_estimation_dataset_params.yaml [hunk 7]
```diff
@@ -27,6 +26,47 @@ joint_links:
...
+ - [31, 119, 180]  # LeftElbow -> LeftArm
+ - [255, 127, 14]  # RightElbow -> RightArm
+ - [148, 103, 189]  # Waist
+ - [255, 127, 14]  # Left Hip -> Left Knee
+ - [214, 39, 40]  # Right Hip -> Right Knee
+ - [31, 119, 180]  # Left Knee -> Left Ankle
+ - [44, 160, 44]  # Right Knee -> Right Ankle
+
+
+keypoint_colors:
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/recipes/dataset_params/coco_pose_estimation_dataset_params.yaml:51]: Keypoint colors are also have specific meaning, in COCO they allow easy differentiation between left and right side of the body

File: src/super_gradients/training/models/pose_estimation_models/dekr_hrnet.py (hunk 1 above)
Comment (BloodAxe) [reply src/super_gradients/training/models/pose_estimation_models/dekr_hrnet.py:567]: @lru_cache(maxsize=1) ?
Copied from detection pipeline. Idea is to cache the pipeline and not re-instantiate it at every call.

File: src/super_gradients/training/models/pose_estimation_models/dekr_hrnet.py (hunk 2 above)
Comment (BloodAxe) [reply src/super_gradients/training/models/pose_estimation_models/dekr_hrnet.py:534]: Not sure I'm following what exactly you are suggesting to move to based class. Can you please elaborate?
From what I read now: 
DEKR model require DEKR-compatible post prediction callback. If we are about to introduce new architecture with different postprocessing we still would have to reimplement this method.

File: src/super_gradients/training/models/prediction_pose_estimation_results.py (hunk 4 above)
Comment (BloodAxe) [reply src/super_gradients/training/models/prediction_pose_estimation_results.py:46]: Yes, we can draw without bounding boxes, once can achieve that by passing `show_confidence=False`.
The goal of drawing bounding box is to help the user understand to which pose this confidence text belongs to.

File: src/super_gradients/training/models/prediction_pose_estimation_results.py (hunk 3 above)
Comment (BloodAxe) [reply src/super_gradients/training/models/prediction_pose_estimation_results.py:15]: Maybe. TBH, I haven't though much about the file location and placed it where the rest of the files related to predict() for object detection were placed:
image

File: src/super_gradients/training/utils/visualization/poses.py (hunk 6 above)
Comment (BloodAxe) [reply src/super_gradients/training/utils/visualization/poses.py]: Renamed to `pose_estimation.py`

File: src/super_gradients/recipes/dataset_params/coco_pose_estimation_dataset_params.yaml [hunk 8]
```diff
@@ -27,6 +26,47 @@ joint_links:
   - [13, 15]
   - [14, 16]
 
+joint_colors:
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/recipes/dataset_params/coco_pose_estimation_dataset_params.yaml]: Agree, edge is better name here. Renamed in e27d373ddbc0a637c8c528483dc9941e4b5589f9

File: src/super_gradients/training/models/prediction_pose_estimation_results.py (hunk 4 above)
Comment (BloodAxe) [reply src/super_gradients/training/models/prediction_pose_estimation_results.py]: However I agree that drawing a box makes a picture cluttered. I changed a default value to False (e.g not showing boxes by default)

With boxes

No boxes

File: src/super_gradients/training/models/prediction_pose_estimation_results.py (hunk 3 above)
Comment (BloodAxe) [reply src/super_gradients/training/models/prediction_pose_estimation_results.py]: Moved to super_gradients.training.utils.predict in 16c29cc93eb8827a32626b4a9f8433a0f5b818ba

File: src/super_gradients/training/models/predictions.py (hunk 5 above)
Comment (BloodAxe) [reply src/super_gradients/training/models/predictions.py]: Moved to super_gradients.training.utils.predict in 16c29cc93eb8827a32626b4a9f8433a0f5b818ba

File: src/super_gradients/training/models/pose_estimation_models/dekr_hrnet.py [hunk 9]
```diff
@@ -520,8 +530,98 @@ def init_weights(self):
...
+        return DEKRPoseEstimationDecodeCallback(
+            min_confidence=conf,
+            keypoint_threshold=0.05,
+            nms_threshold=0.05,
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/models/pose_estimation_models/dekr_hrnet.py:538]: We provide the control over the most top-level one `conf`. Yes, we can export all parameters to give user full control. 
But without deep knowledge of what each parameter influence to that would be a random guessing for the user. 

There are few problems with exposing all parameters:
1) We would have to duplicate them in 6 (six) places: in `draw` and `show` calls for three classes: `ImagePoseEstimationPredictions`, `ImagesPoseEstimationPredictions`, `VideoPoseEstimationPredictions`.
2) Most of these hyperparams are specific to DEKR postprocessing, while other architectures (one day) may have another set of parameters. Will we have to include them all as well? 
WDYT?

File: src/super_gradients/training/models/predictions.py [hunk 10]
```diff
@@ -1,55 +0,0 @@
-from typing import Tuple
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/models/predictions.py:1]: These are not public API (In a sense we do not import it directly anywhere, to my knowledge), so this should be safe to just move them and change imports internally.
Notebooks do not import that stuff. Correct me if I'm wrong.

File: src/super_gradients/training/datasets/pose_estimation_datasets/coco_keypoints.py [hunk 11]
```diff
@@ -49,8 +52,17 @@ def __init__(
...
+        :param edge_links: List of joint links to be visualized on the image
+        :param edge_colors: List of colors for each joint link
         """
-        super().__init__(transforms=transforms, target_generator=target_generator, min_instance_area=min_instance_area)
+        super().__init__(
+            transforms=transforms,
+            target_generator=target_generator,
+            min_instance_area=min_instance_area,
+            edge_links=edge_links,
+            edge_colors=edge_colors,
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/datasets/pose_estimation_datasets/coco_keypoints.py:73]: Added in 6d914b5c8eeb36f4d226dc7dc6d580a77331e562

File: src/super_gradients/training/models/predictions.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/models/predictions.py:1]: Low risk, but still possible.
We can take our chances...up to you.

File: src/super_gradients/training/models/pose_estimation_models/dekr_hrnet.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/models/pose_estimation_models/dekr_hrnet.py:538]: I am not that familiar with pose NMS, but it seemed to me that we should make nms_threshold controllable.
Digging in that led me to:
super_gradients.training.utils.pose_estimation.dekr_decode_callbacks.pose_nms

Which holds docstring which is not up-todate.
Also, super_gradients.training.utils.pose_estimation.dekr_decode_callbacks.DEKRPoseEstimationDecodeCallback has some docs missing,

File: src/super_gradients/training/models/predictions.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/models/predictions.py:1]: You're actually right, better safe than sorry.
Added re-import with deprecation warning in old file.

**Review by shaydeci — APPROVED** [review]
LGTM

**Pull request merged**